test/wasm_test.cc

```diff
@@ -113,4 +114,132 @@ TEST_P(TestVm, GetOrCreateThreadLocalWasmFailCallbacks) {
+  EXPECT_EQ(canary_count, 1);
+
+  std::unordered_set<std::shared_ptr<WasmHandleBase>> reference_holder;
+
+  for (const auto &root_id : root_ids) {
+    for (const auto &vm_id : vm_ids) {
+      for (const auto &vm_config : vm_configs) {
+        for (const auto &plugin_key : plugin_keys) {
+          for (const auto &plugin_config : plugin_configs) {
+            canary_count = 0;
+            TestContext::resetGlobalLog();
+            WasmHandleFactory wasm_handle_factory_comp =
+                [this, vm_id,
+                 vm_config](std::string_view vm_key) -> std::shared_ptr<WasmHandleBase> {
+              auto base_wasm = std::make_shared<TestWasm>(
+                  newVm(), std::unordered_map<std::string, std::string>(), vm_id, vm_config,
+                  vm_key);
+              return std::make_shared<WasmHandleBase>(base_wasm);
+            };
+            const auto plugin_comp = std::make_shared<PluginBase>(
+                plugin_name, root_id, vm_id, engine_, plugin_config, fail_open, plugin_key);
+            const auto vm_key = makeVmKey(vm_id, vm_config, "common_code");
+            // Create a base Wasm by createWasm.
+            auto wasm_handle_comp =
+                createWasm(vm_key, source, plugin_comp, wasm_handle_factory_comp,
+                           wasm_handle_clone_factory_for_canary, false);
+
+            if (plugin_config.empty()) {
+              // configure_check.wasm should raise the error at `onConfigure` in canary when the
+              // `plugin_config` is empty string.
+              EXPECT_EQ(wasm_handle_comp, nullptr);
+              continue;
+            }
+
+            ASSERT_TRUE(wasm_handle_comp && wasm_handle_comp->wasm());
+            // Keep the reference of wasm_handle_comp in order to utilize the WasmHandleBase
+            // cache of createWasm. If we don't keep the reference, WasmHandleBase and VM will be
+            // destroyed for each iteration.
+            reference_holder.insert(wasm_handle_comp);
+
+            EXPECT_TRUE(TestContext::isGlobalLogged("onConfigure: " + root_id));
+
+            // Wasm VM is unique for vm_key.
+            if (vm_key == vm_key_baseline) {
+              EXPECT_EQ(wasm_handle_baseline->wasm(), wasm_handle_comp->wasm());
+            } else {
+              EXPECT_NE(wasm_handle_baseline->wasm(), wasm_handle_comp->wasm());
+            }
+
+            // plugin->key() is unique for root_id + plugin_config + plugin_key.
+            // plugin->key() is used as an identifier of local-specific plugins as well.
+            if (root_id == root_ids[0] && plugin_config == plugin_configs[0] &&
+                plugin_key == plugin_keys[0]) {
+              EXPECT_EQ(plugin_baseline->key(), plugin_comp->key());
+            } else {
+              EXPECT_NE(plugin_baseline->key(), plugin_comp->key());
+            }
+            // For each create Wasm, canary should be done.
+            EXPECT_EQ(canary_count, 1);
```

Comment:
I believe `canary_count` is increased with each clone, regardless of whether config canary was successful or not, so you might want to check it right after `createWasm`, or verify that `canary_count` matches total number of iterations. Checking it here doesn't really prove anything.

Reply:
Moved to right after of the `creatWasm`.